Cabal/Distribution/PackageDescription/Check.hs

```diff
@@ -253,6 +253,18 @@ checkLibrary pkg lib =
            "An 'autogen-module' is neither on 'exposed-modules' or "
         ++ "'other-modules'."
 
+    -- check that all autogen-c-sources appear on c-sources
+  , check
+      (not $ and $ map (flip elem
+        (cSources $ libBuildInfo lib)) (autogenCSources $ libBuildInfo lib)) $
+      PackageBuildImpossible $ "An 'autogen-c-sources' is not on 'c-sources'"
```

Comment:
1) The real reason why we repeat filenames is not for backward compatibility is to be consistent with how `autogen-modules` work, it would be awkward to explain that on an error message.

2) Do you mean a helper function inside this module that does `not $ and $ map (flip elem e)` or something like libModules and `libModulesAutogen` on `Distribution.Types.Library` ?